flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/ExecutionGraphDeploymentTest.java

```diff
@@ -662,6 +674,91 @@ public void testEagerSchedulingWaitsOnAllInputPreferredLocations() throws Except
 		}
 	}
 
+	/**
+	 * Tests that the {@link ExecutionGraph} is deployed in topological order.
+	 */
+	@Test
+	public void testExecutionGraphIsDeployedInTopologicalOrder() throws Exception {
+		final int sourceParallelism = 2;
+		final int sinkParallelism = 1;
+
+		final JobVertex sourceVertex = new JobVertex("source");
+		sourceVertex.setInvokableClass(NoOpInvokable.class);
+		sourceVertex.setParallelism(sourceParallelism);
+
+		final JobVertex sinkVertex = new JobVertex("sink");
+		sinkVertex.setInvokableClass(NoOpInvokable.class);
+		sinkVertex.setParallelism(sinkParallelism);
+
+		sinkVertex.connectNewDataSetAsInput(sourceVertex, DistributionPattern.POINTWISE, ResultPartitionType.PIPELINED);
+
+		final JobID jobId = new JobID();
+		final int numberTasks = sourceParallelism + sinkParallelism;
+		final ArrayBlockingQueue<ExecutionAttemptID> submittedTasksQueue = new ArrayBlockingQueue<>(numberTasks);
+		TestingTaskExecutorGatewayBuilder testingTaskExecutorGatewayBuilder = new TestingTaskExecutorGatewayBuilder();
+		testingTaskExecutorGatewayBuilder.setSubmitTaskConsumer((taskDeploymentDescriptor, jobMasterId) -> {
+			submittedTasksQueue.offer(taskDeploymentDescriptor.getExecutionAttemptId());
+			return CompletableFuture.completedFuture(Acknowledge.get());
+		});
+
+		final TestingTaskExecutorGateway taskExecutorGateway = testingTaskExecutorGatewayBuilder.createTestingTaskExecutorGateway();
+		final RpcTaskManagerGateway taskManagerGateway = new RpcTaskManagerGateway(taskExecutorGateway, JobMasterId.generate());
+
+		final Collection<CompletableFuture<LogicalSlot>> slotFutures = new ArrayList<>(numberTasks);
+		for (int i = 0; i < numberTasks; i++) {
+			slotFutures.add(new CompletableFuture<>());
+		}
+
+		final Iterator<CompletableFuture<LogicalSlot>> slotIterator = slotFutures.iterator();
+		final SlotProvider slotProvider = new TestingSlotProvider(slotRequestId -> {
```

Comment:
This looks like a basic use-case that we could move into a reusable `FixedSizeTestingSlotProvider extends TestingSlotProvider` class with a `Collection<CompletableFuture<LogicalSlot>> getSlotFutures()` method.

Reply:
Good point. I will add this utility and refactor the test.